patches/api/0420-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,746 @@
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..50a54e913a2aad1dcf117814de353e9b6fcadcc6
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java
+@@ -0,0 +1,87 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.mojang.brigadier.StringReader;
++import com.mojang.brigadier.arguments.ArgumentType;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++
++/**
++ * An argument type that wraps around another argument.
++ * This argument is special in that the argument type that is wrapped will
++ * be sent to the client, not the wrapper itself.
++ *
++ * @param <T> type
++ * @param <B> base wrapping type
++ */
++@ApiStatus.Experimental
++public abstract class CustomArgumentType<T, B> implements ArgumentType<T> {
++
++    @NotNull
++    private final ArgumentType<B> vanillaPrimitiveType;
++
++    public CustomArgumentType(@NotNull ArgumentType<B> primitive) {
++        VanillaArguments.PROVIDER.validatePrimitiveType(primitive);
++        this.vanillaPrimitiveType = primitive;
++    }
++
++    /**
++     * Parses the argument using the wrapped argument type then converting
++     * it.
++     * @param reader string reader
++     * @return value
++     * @throws CommandSyntaxException if an error occurs while parsing
++     */
++    @NotNull
++    @Override
++    public T parse(@NotNull StringReader reader) throws CommandSyntaxException {
++        return this.convert(this.vanillaPrimitiveType.parse(reader));
++    }
++
++    /**
++     * Converts the type from the wrapped argument to this argument.
++     *
++     * @param baseType wrapped argument provided value
++     * @return converted value
++     * @throws CommandSyntaxException if an exception occurs while parsing
++     */
++    @NotNull
++    public abstract T convert(@NotNull B baseType) throws CommandSyntaxException;
++
++    /**
++     * Gets the primitive type that this argument uses,
++     * or the type that is sent to the client.
++     *
++     * @return primitive argument
++     */
++    @NotNull
++    public final ArgumentType<B> getPrimitiveType() {
++        return this.vanillaPrimitiveType;
++    }
++
++    /**
++     * Cannot be controlled by the server.
++     * Returned in cases where there are multiple arguments in the same node.
++     * This helps differentiate and tell the player what the possible inputs are.
++     *
++     * @return client set examples
++     */
++    @NotNull
++    @Override
++    public final Collection<String> getExamples() {
++        return this.vanillaPrimitiveType.getExamples();
```

Comment:
The examples are used in the built-in validation/ambiguities system. Is it worth allowing plugins to take advantage of that, and then running the validate to check after commands have been added by a plugin?

Reply:
Am not sure, I feel like it would complicate things.

Reply:
Well it would help detect command conflicts right? Like if 2 plugins tried to add commands at the same node.

Reply:
I guess the issue arises that this will introduce possible confusion and desync between the client's suggestions vs the server. So, not sure if we should encourage changing this.